src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,341 @@
+
+                    // 1.3. After processing the packet header, there is a possibility that the transport read didn't
+                    // return any more data for the packet. In that case, post another read to have packet data ready..
+                    if (_readBufferDataEnd == _readIndex)
+                    {
+                        _readBufferDataEnd = isAsync ?
+                            await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                            _underlyingStream.Read(_readBuffer);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// If the buffer is empty, or the packet data is completely read, 
+        /// then we need to prepare the buffer for the next packet.
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async ValueTask PrepareBufferIfNeeded(bool isAsync, CancellationToken ct)
+        {
+            // If there we have read through the packet, 
+            // or if we have reached the end of the buffer,
+            // then we need to position the buffer at the beginning of packet or buffer
+            if (_packetDataLeft == 0 || _readBufferDataEnd == _readIndex)
+                await PrepareBufferAsync(isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Processes the header of the packet, and extracts the data from the header.
+        /// If needed, this function will read more data from the network to complete the header.
+        /// </summary>
+        /// <param name="isAsync">Whether this method is invoked asynchronously.</param>
+        /// <param name="ct">The cancellation token.</param>
+        /// <returns></returns>
+        private async ValueTask ProcessHeaderAsync(bool isAsync, CancellationToken ct)
+        {
+            int headerDataAvailable = _readBufferDataEnd - _readIndex;
+            int bytesNeededToCompleteHeader = TdsEnums.HEADER_LEN - headerDataAvailable;
+
+            // We have less than the header length available in the buffer, so we need to read more data, to atleast complete 
+            // the header.
+            if (headerDataAvailable < TdsEnums.HEADER_LEN)
+            {
+                // We move the header information to the beginning of the buffer.
+                Buffer.BlockCopy(_readBuffer, _readIndex, _readBuffer, 0, headerDataAvailable);
+                _readBufferDataEnd = headerDataAvailable;
+                _readIndex = 0;
+
+                while (bytesNeededToCompleteHeader > 0)
+                {
+                    int bytesRead = isAsync ? 
+                        await _underlyingStream.ReadAsync(_readBuffer.AsMemory(_readBufferDataEnd), ct).ConfigureAwait(false) 
+                            : _underlyingStream.Read(_readBuffer, _readBufferDataEnd, bytesNeededToCompleteHeader);
```

Comment:
```suggestion
                            : _underlyingStream.Read(_readBuffer.AsSpan(_readBufferDataEnd));
```

If more data than expected is available, this'll allow that data to be copied into the buffer, which should reduce the need for case 1.3 in `PrepareBufferAsync`.